sdk/storage/Azure.Storage.Queues/api/Azure.Storage.Queues.netstandard2.0.cs

```diff
@@ -315,6 +315,27 @@ public partial class UpdateReceipt
         public string PopReceipt { get { throw null; } }
     }
 }
+namespace Azure.Storage.Queues.Specialized
+{
+    public partial interface IClientSideDecryptionFailureListener
+    {
+        void OnFailure(Azure.Storage.Queues.Models.PeekedMessage message, System.Exception exception);
+        void OnFailure(Azure.Storage.Queues.Models.QueueMessage message, System.Exception exception);
+        System.Threading.Tasks.Task OnFailureAsync(Azure.Storage.Queues.Models.PeekedMessage message, System.Exception exception);
+        System.Threading.Tasks.Task OnFailureAsync(Azure.Storage.Queues.Models.QueueMessage message, System.Exception exception);
+    }
```

Comment:
If we believe we don't need async methods here then so be it. But we have two distinct types of QueueMessage and PeekedMessage that don't have a common base. Is accepting more than one action when the user probably wants to always supply both really the best option? I'm not sure of a better way to supply this functionality.

Reply:
@KrzysztofCwalina and I talked about this a bit and came up with the following to match what we did for QuickQuery a little better here:

```C#
    public class QueueClientSideEncryptionOptions : ClientSideEncryptionOptions
    {
        public event EventHandler<DecryptionFailureEventArgs> DecryptionFailed;
    }

    public class ClientSideDecryptionFailureEventArgs
    {
        public Exception Exception { get; }
        public ClientSideDecryptionFailureEventArgs(Exception exception);
    }
```

and you could use it like:

```C#
var options = new QueueClientSideEncryptionOptions();
options.KeyResolver = ...;
options.DecryptionFailed +=
    (object sender, ClientSideDecryptionFailureEventArgs args) =>
    {
        if (sender is PeekedMessage peek)
        {
             Console.WriteLine($"Failed to peek at {peek.MessageId}");
        }
        else if (sender is QueueMessage message)
        {
             Console.WriteLine($"Failed to receive {message.MessageId}");
        }
    };
```

It's only sync for now, but so is QuickQuery's error handler and I'd wait to hear feedback from folks if they need an async event.

Do you this this would work?

Reply:
Is the suggestion that `SpecializedQueueClientOptions` now have a property for `QueueClientSideEncryptionOptions` instead of the normal `ClientSideEncryptionOptions`, or that we still accept the base class and work it through internals like we do with client options in the first place?

Reply:
I'd say leave it as the base `ClientSideEncryptionOptions` in `SpecializedQueueClientOptions ` so people could pass the same value for Blobs and Queues.  You'd only need to learn about `QueueClientSideEncryptionOptions` if you were trying to handle failures which should be an advanced scenario.

On a related note - maybe we should change the semantics to throw an exception when you can't decrypt if there's no failure handler?  Then new users who aren't trying to use multiple keys/etc. will still have an easy notification that something went wrong but advanced scenarios are still possible for special partners?

Reply:
And you could just do an `if (options is QueueClientSideEncryptionOptions advanced) { _failureHandler = advanced.DecryptionFailed; }` instead of a bunch of internal state sharing since `ClientSideEncryptionOptions` is in Common.